File: 2022/day_07.rs
use std::cell::RefCell;
use std::collections::{HashMap, VecDeque};
use std::rc::Rc;

#[derive(Debug, Default, Eq, PartialEq)]
enum FileKind {
    #[default]
    File,
    Dir,
}

#[derive(Debug, Default)]
struct File {
    kind: FileKind,
    size: usize,
    parent: Option<Rc<RefCell<File>>>,
    children: HashMap<String, Rc<RefCell<File>>>,
}

fn main() {
    let data = std::fs::read_to_string("input.txt").unwrap();
    dbg!(sum_of_sizes(data));
}

fn sum_of_sizes(data: String) -> usize {
    let mut files = VecDeque::new();
    let mut indegree = std::collections::HashMap::new();
    let root = Rc::new(RefCell::new(File {
        kind: FileKind::Dir,
        size: 0,
        parent: None,
        children: HashMap::new(),
    }));
    let mut node = Rc::clone(&root);
    for line in data.split('\n') {
        let tokens = line.split(' ').collect::<Vec<_>>();
        match tokens[0] {
            "$" => {
                if tokens[1] != "cd" {
                    continue;
                }
                node = match tokens[2] {
                    "/" => Rc::clone(&root),
                    ".." => {
                        if let Some(ref par) = node.borrow().parent {
                            Rc::clone(par)
                        } else {
                            Rc::clone(&root)
                        }
                    }
                    file => Rc::clone(&node.borrow().children[file]),
                };
            }
            "dir" => {
                node.borrow_mut().children.insert(
                    String::from(tokens[1]),
                    Rc::new(RefCell::new(File {
                        kind: FileKind::Dir,
                        size: 0,
                        parent: Some(Rc::clone(&node)),
                        children: HashMap::new(),
                    })),
                );
                *indegree.entry(node.as_ptr()).or_insert(0) += 1;
            }
            "" => {}
            size => {
                let file = Rc::new(RefCell::new(File {
                    kind: FileKind::File,
                    size: size.parse::<usize>().unwrap(),
                    parent: Some(Rc::clone(&node)),
                    children: HashMap::new(),
                }));
                node.borrow_mut()
                    .children
                    .insert(String::from(tokens[1]), Rc::clone(&file));
                *indegree.entry(node.as_ptr()).or_insert(0) += 1;
                files.push_back(file);
            }
        }
    }
    let mut sizes = vec![0];
    while let Some(file) = files.pop_front() {
        let file = file.borrow();
        if let Some(ref par) = file.parent {
            par.borrow_mut().size += file.size;
            *indegree.get_mut(&par.as_ptr()).unwrap() -= 1;
            if indegree[&par.as_ptr()] == 0 {
                files.push_back(Rc::clone(par));
            }
        }
        if file.kind == FileKind::Dir {
            sizes.push(file.size);
        }
    }
    sizes.sort();
    const REQUIRED: usize = 30000000;
    let unused_space = 70000000 - sizes[sizes.len() - 1];
    for x in sizes {
        if unused_space + x >= REQUIRED {
            return x;
        }
    }
    unreachable!()
}

File: 2022/day_07.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

type File struct {
	kind     int // 0: Dir, 1: File
	size     int
	parent   *File
	children map[string]*File
}

func create_dir(parent *File) *File {
	return &File{
		kind:     0,
		parent:   parent,
		children: map[string]*File{},
	}
}

func solve(scanner *bufio.Scanner) {
	files := []*File{}
	indegree := map[*File]int{}
	root := create_dir(nil)
	root.parent = root
	node := root
	for scanner.Scan() {
		tokens := strings.Fields(scanner.Text())
		switch tokens[0] {
		case "$":
			switch tokens[1] {
			case "cd":
				switch tokens[2] {
				case "/":
					node = root
				case "..":
					node = node.parent
				default:
					node = node.children[tokens[2]]
				}
			}
		case "dir":
			node.children[tokens[1]] = create_dir(node)
			indegree[node] += 1
		default:
			node.children[tokens[1]] = &File{
				kind:   1,
				size:   str_to_int(tokens[0]),
				parent: node,
			}
			indegree[node] += 1
			files = append(files, node.children[tokens[1]])
		}
	}
	sizes := []int{0}
	for i := 0; i < len(files); i++ {
		file := files[i]
		parent := file.parent
		if parent != file {
			parent.size += file.size
			indegree[parent] -= 1
			if indegree[parent] == 0 {
				files = append(files, parent)
			}
		}
		if file.kind == 0 {
			sizes = append(sizes, file.size)
		}
	}
	sort.Ints(sizes)
	remaining := 70000000 - sizes[len(sizes)-1]
	required := 30000000
	for _, size := range sizes {
		if size+remaining >= required {
			fmt.Println(size)
			return
		}
	}
}

func str_to_int(end string) int {
	v, _ := strconv.ParseInt(end, 10, 64)
	return int(v)
}

var input_files = [2]string{"test.txt", "input.txt"}

func main() {
	for _, input_file := range input_files {
		file, err := os.Open(input_file)
		if err != nil {
			panic(err)
		}
		defer file.Close()
		scanner := bufio.NewScanner(file)
		solve(scanner)
	}
}
